File: cpp/test/hsm_machine_test.cpp
#include <vector>

#include <gtest/gtest.h>

#include "hsm.hpp"

namespace {

struct RecordingHooks final : statesurf::IHooks {
  std::vector<statesurf::State> entries;
  std::vector<statesurf::State> exits;
  std::vector<statesurf::ActionId> actions;
  std::vector<statesurf::GuardId> guard_calls;
  bool foo = true;

  void on_entry(statesurf::State s) override { entries.push_back(s); }
  void on_exit(statesurf::State s) override { exits.push_back(s); }

  bool guard(statesurf::State, statesurf::Event, statesurf::GuardId id) override {
    guard_calls.push_back(id);
    switch (id) {
      case statesurf::GuardId::isFooTrue:
        return foo;
      case statesurf::GuardId::isFooFalse:
        return !foo;
      default:
        return false;
    }
  }

  void action(statesurf::State, statesurf::Event, statesurf::ActionId id) override {
    actions.push_back(id);
    switch (id) {
      case statesurf::ActionId::setFooFalse:
        foo = false;
        break;
      case statesurf::ActionId::setFooTrue:
        foo = true;
        break;
      default:
        break;
    }
  }

  void reset_logs() {
    entries.clear();
    exits.clear();
    actions.clear();
    guard_calls.clear();
  }
};

}  // namespace

TEST(StateSurfMachine, DrivesThroughLifecycle) {
  RecordingHooks hooks;
  statesurf::StateSurfMachine machine(hooks);

  EXPECT_EQ(machine.state(), statesurf::State::StateSurfInitial);
  EXPECT_FALSE(machine.terminated());
  EXPECT_TRUE(hooks.entries.empty());
  EXPECT_TRUE(hooks.actions.empty());

  machine.start();

  const std::vector<statesurf::State> expected_initial_entries{
      statesurf::State::s,
      statesurf::State::s2,
      statesurf::State::s21,
      statesurf::State::s211};
  EXPECT_EQ(hooks.entries, expected_initial_entries);
  EXPECT_TRUE(hooks.exits.empty());
  const std::vector<statesurf::ActionId> expected_initial_actions{
      statesurf::ActionId::setFooFalse};
  EXPECT_EQ(hooks.actions, expected_initial_actions);
  EXPECT_FALSE(hooks.foo);
  EXPECT_FALSE(machine.terminated());
  EXPECT_EQ(machine.state(), statesurf::State::s211);

  hooks.reset_logs();

  auto dispatch_and_expect = [&](statesurf::Event event,
                                 std::vector<statesurf::State> expected_exits,
                                 std::vector<statesurf::State> expected_entries,
                                 std::vector<statesurf::ActionId> expected_actions,
                                 std::vector<statesurf::GuardId> expected_guards,
                                 statesurf::State expected_state) {
    machine.dispatch(event);
    EXPECT_EQ(hooks.exits, expected_exits);
    EXPECT_EQ(hooks.entries, expected_entries);
    EXPECT_EQ(hooks.actions, expected_actions);
    EXPECT_EQ(hooks.guard_calls, expected_guards);
    EXPECT_EQ(machine.state(), expected_state);
    EXPECT_FALSE(machine.terminated());
    hooks.reset_logs();
  };

  dispatch_and_expect(
      statesurf::Event::G,
      {statesurf::State::s211, statesurf::State::s21, statesurf::State::s2},
      {statesurf::State::s1, statesurf::State::s11},
      {},
      {},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::I,
      {},
      {},
      {},
      {},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::A,
      {statesurf::State::s11},
      {statesurf::State::s1, statesurf::State::s11},
      {},
      {},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::D,
      {statesurf::State::s11},
      {statesurf::State::s11},
      {statesurf::ActionId::setFooTrue},
      {statesurf::GuardId::isFooTrue, statesurf::GuardId::isFooFalse},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::D,
      {},
      {},
      {statesurf::ActionId::setFooFalse},
      {statesurf::GuardId::isFooTrue},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::C,
      {statesurf::State::s11, statesurf::State::s1},
      {statesurf::State::s2, statesurf::State::s21, statesurf::State::s211},
      {},
      {},
      statesurf::State::s211);

  dispatch_and_expect(
      statesurf::Event::E,
      {statesurf::State::s211, statesurf::State::s21, statesurf::State::s2},
      {statesurf::State::s1, statesurf::State::s11},
      {},
      {},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::E,
      {statesurf::State::s11, statesurf::State::s1},
      {statesurf::State::s1, statesurf::State::s11},
      {},
      {},
      statesurf::State::s11);

  dispatch_and_expect(
      statesurf::Event::G,
      {statesurf::State::s11, statesurf::State::s1},
      {statesurf::State::s2, statesurf::State::s21, statesurf::State::s211},
      {},
      {},
      statesurf::State::s211);

  dispatch_and_expect(
      statesurf::Event::I,
      {},
      {},
      {statesurf::ActionId::setFooTrue},
      {statesurf::GuardId::isFooFalse},
      statesurf::State::s211);

  dispatch_and_expect(
      statesurf::Event::I,
      {},
      {},
      {statesurf::ActionId::setFooFalse},
      {statesurf::GuardId::isFooFalse, statesurf::GuardId::isFooTrue},
      statesurf::State::s211);

  machine.dispatch(statesurf::Event::TERMINATE);
  EXPECT_TRUE(machine.terminated());
  EXPECT_EQ(machine.state(), statesurf::State::StateSurfFinal);
}

File: rust/tests/hsm_machine_test.rs
use state_surf::generated::hsm::hsm::{
    HsmActionId, HsmCallbacks, HsmEvent, HsmGuardId, HsmMachine, HsmState,
};

struct RecordingCallbacks {
    entries: Vec<HsmState>,
    exits: Vec<HsmState>,
    actions: Vec<HsmActionId>,
    guard_calls: Vec<HsmGuardId>,
    foo: bool,
}

impl Default for RecordingCallbacks {
    fn default() -> Self {
        Self::new()
    }
}

impl RecordingCallbacks {
    fn new() -> Self {
        Self {
            entries: Vec::new(),
            exits: Vec::new(),
            actions: Vec::new(),
            guard_calls: Vec::new(),
            foo: true,
        }
    }

    fn reset_logs(&mut self) {
        self.entries.clear();
        self.exits.clear();
        self.actions.clear();
        self.guard_calls.clear();
    }
}

impl HsmCallbacks for RecordingCallbacks {
    fn on_entry(&mut self, state: HsmState) {
        self.entries.push(state);
    }

    fn on_exit(&mut self, state: HsmState) {
        self.exits.push(state);
    }

    fn guard(&mut self, _state: HsmState, _event: HsmEvent, guard: HsmGuardId) -> bool {
        self.guard_calls.push(guard);
        match guard {
            HsmGuardId::isFooTrue => self.foo,
            HsmGuardId::isFooFalse => !self.foo,
        }
    }

    fn action(&mut self, _state: HsmState, _event: HsmEvent, action: HsmActionId) {
        self.actions.push(action);
        match action {
            HsmActionId::setFooFalse => self.foo = false,
            HsmActionId::setFooTrue => self.foo = true,
        }
    }
}

fn dispatch_and_expect(
    machine: &mut HsmMachine<RecordingCallbacks>,
    event: HsmEvent,
    expected_exits: &[HsmState],
    expected_entries: &[HsmState],
    expected_actions: &[HsmActionId],
    expected_guards: &[HsmGuardId],
    expected_state: HsmState,
) {
    machine.dispatch(event);
    {
        let callbacks = machine.callbacks();
        assert_eq!(callbacks.exits.as_slice(), expected_exits);
        assert_eq!(callbacks.entries.as_slice(), expected_entries);
        assert_eq!(callbacks.actions.as_slice(), expected_actions);
        assert_eq!(callbacks.guard_calls.as_slice(), expected_guards);
    }
    assert_eq!(machine.state(), expected_state);
    assert!(!machine.terminated());
    machine.callbacks_mut().reset_logs();
}

#[test]
fn drives_through_lifecycle() {
    let callbacks = RecordingCallbacks::new();
    let mut machine = HsmMachine::new(callbacks);

    {
        let callbacks_view = machine.callbacks();
        assert_eq!(machine.state(), HsmState::InitialPseudoState);
        assert!(!machine.terminated());
        assert!(callbacks_view.entries.is_empty());
        assert!(callbacks_view.actions.is_empty());
        assert!(callbacks_view.guard_calls.is_empty());
    }

    machine.start();
    {
        let callbacks_view = machine.callbacks();
        let expected_initial_entries = vec![HsmState::s, HsmState::s2, HsmState::s21, HsmState::s211];
        assert_eq!(callbacks_view.entries, expected_initial_entries);
        assert!(callbacks_view.exits.is_empty());
        let expected_initial_actions = vec![HsmActionId::setFooFalse];
        assert_eq!(callbacks_view.actions, expected_initial_actions);
        assert!(callbacks_view.guard_calls.is_empty());
        assert!(!callbacks_view.foo);
    }
    assert!(!machine.terminated());
    assert_eq!(machine.state(), HsmState::s211);
    machine.callbacks_mut().reset_logs();

    dispatch_and_expect(
        &mut machine,
        HsmEvent::G,
        &[HsmState::s211, HsmState::s21, HsmState::s2],
        &[HsmState::s1, HsmState::s11],
        &[],
        &[],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::I,
        &[],
        &[],
        &[],
        &[],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::A,
        &[HsmState::s11, HsmState::s1],
        &[HsmState::s1, HsmState::s11],
        &[],
        &[],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::D,
        &[HsmState::s11, HsmState::s1],
        &[HsmState::s1, HsmState::s11],
        &[HsmActionId::setFooTrue],
        &[HsmGuardId::isFooTrue, HsmGuardId::isFooFalse],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::D,
        &[HsmState::s11],
        &[HsmState::s11],
        &[HsmActionId::setFooFalse],
        &[HsmGuardId::isFooTrue],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::C,
        &[HsmState::s11, HsmState::s1],
        &[HsmState::s2, HsmState::s21, HsmState::s211],
        &[],
        &[],
        HsmState::s211,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::E,
        &[HsmState::s211, HsmState::s21, HsmState::s2],
        &[HsmState::s1, HsmState::s11],
        &[],
        &[],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::E,
        &[HsmState::s11, HsmState::s1],
        &[HsmState::s1, HsmState::s11],
        &[],
        &[],
        HsmState::s11,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::G,
        &[HsmState::s11, HsmState::s1],
        &[HsmState::s2, HsmState::s21, HsmState::s211],
        &[],
        &[],
        HsmState::s211,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::I,
        &[],
        &[],
        &[HsmActionId::setFooTrue],
        &[HsmGuardId::isFooFalse],
        HsmState::s211,
    );

    dispatch_and_expect(
        &mut machine,
        HsmEvent::I,
        &[],
        &[],
        &[HsmActionId::setFooFalse],
        &[HsmGuardId::isFooFalse, HsmGuardId::isFooTrue],
        HsmState::s211,
    );

    machine.dispatch(HsmEvent::TERMINATE);
    assert!(machine.terminated());
    assert_eq!(machine.state(), HsmState::FinalPseudoState);
    {
        let callbacks_view = machine.callbacks();
        assert_eq!(callbacks_view.exits, vec![HsmState::s211, HsmState::s21, HsmState::s2, HsmState::s]);
        assert_eq!(callbacks_view.entries, vec![HsmState::FinalPseudoState]);
    }
}
